File: src/main/java/com/bajdcc/LALR1/syntax/lexer/tokenizer/OperatorTokenizer.java
package com.bajdcc.LALR1.syntax.lexer.tokenizer;

import com.bajdcc.LALR1.syntax.token.OperatorType;
import com.bajdcc.util.lexer.algorithm.TokenAlgorithm;
import com.bajdcc.util.lexer.error.RegexException;
import com.bajdcc.util.lexer.regex.IRegexStringIterator;
import com.bajdcc.util.lexer.token.MetaType;
import com.bajdcc.util.lexer.token.Token;
import com.bajdcc.util.lexer.token.TokenType;

import java.util.HashMap;

/**
 * 操作符解析
 *
 * @author bajdcc
 */
public class OperatorTokenizer extends TokenAlgorithm {

	/**
	 * 关键字哈希表
	 */
	private HashMap<String, OperatorType> hashOperator = new HashMap<>();

	public OperatorTokenizer() throws RegexException {
		super(getRegexString(), null);
		initializeHashMap();
	}

	public static String getRegexString() {
		String exp = "->|\\||(|)|[|]|{|}|<|>";
		MetaType[] metaTypes = new MetaType[]{MetaType.LPARAN,
				MetaType.RPARAN, MetaType.STAR, MetaType.PLUS,
				MetaType.LSQUARE, MetaType.RSQUARE, MetaType.LBRACE,
				MetaType.RBRACE, MetaType.DOT};
		for (MetaType meta : metaTypes) {
			exp = exp.replace(meta.getChar() + "", "\\" + meta.getChar());
		}
		return exp;
	}

	@Override
	public boolean getGreedMode() {
		return true;
	}

	/**
	 * 初始化关键字哈希表
	 */
	private void initializeHashMap() {
		for (OperatorType operator : OperatorType.values()) {// 关键字
			hashOperator.put(operator.getName(), operator);
		}
	}

	/*
	 * （非 Javadoc）
	 *
	 * @see
	 * com.bajdcc.lexer.algorithm.ITokenAlgorithm#getToken(java.lang.String,
	 * com.bajdcc.lexer.token.Token)
	 */
	@Override
	public Token getToken(String string, Token token, IRegexStringIterator iterator) {
		token.kToken = TokenType.OPERATOR;
		token.object = hashOperator.get(string);
		return token;
	}
}

File: src/main/kotlin/com/bajdcc/LALR1/syntax/lexer/tokenizer/OperatorTokenizer.kt
package com.bajdcc.LALR1.syntax.lexer.tokenizer

import com.bajdcc.LALR1.syntax.token.OperatorType
import com.bajdcc.util.lexer.algorithm.TokenAlgorithm
import com.bajdcc.util.lexer.error.RegexException
import com.bajdcc.util.lexer.regex.IRegexStringIterator
import com.bajdcc.util.lexer.token.MetaType
import com.bajdcc.util.lexer.token.Token
import com.bajdcc.util.lexer.token.TokenType

/**
 * 操作符解析
 *
 * @author bajdcc
 */
class OperatorTokenizer @Throws(RegexException::class)
constructor() : TokenAlgorithm(regexString, null) {

    override val greedMode: Boolean
        get() = true

    override fun getToken(string: String, token: Token, iterator: IRegexStringIterator): Token? {
        token.type = TokenType.OPERATOR
        token.obj = hashOperator[string]
        return token
    }

    companion object {

        /**
         * 关键字哈希表
         */
        private val hashOperator = mutableMapOf<String, OperatorType>()

        init {
            for (operator in OperatorType.values()) {// 关键字
                hashOperator[operator.desc] = operator
            }
        }

        val regexString: String
            get() {
                var exp = "->|\\||(|)|[|]|{|}|<|>"
                val metaTypes = arrayOf(MetaType.LPARAN, MetaType.RPARAN, MetaType.STAR, MetaType.PLUS, MetaType.LSQUARE, MetaType.RSQUARE, MetaType.LBRACE, MetaType.RBRACE, MetaType.DOT)
                for (meta in metaTypes) {
                    exp = exp.replace(meta.char + "", "\\" + meta.char)
                }
                return exp
            }
    }
}
